### PR TITLE
[corefoundation] Add missing CFStringTransform API

### PR DESCRIPTION
**Pull request by spouliot**
Since there's not NSString or .NET equivalent. This requires a mutable
`CFString` so the new type is also added.

Draft until unit tests are added.

### CODE REVIEW

File: src/CoreFoundation/CFMutableString.cs [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+		static extern internal bool /* Boolean */ CFStringTransform (/* CFMutableStringRef* */ IntPtr @string, /* CFRange* */ ref CFRange range, /* CFStringRef* */ IntPtr transform, [MarshalAs (UnmanagedType.I1)] /* Boolean */ bool reverse);
+
+		public bool Transform (ref CFRange range, CFStringTransform transform, bool reverse)
+		{
+			str = null; // destroy any cached value
+			return Transform (Handle, ref range, transform.GetConstant ().GetHandle (), reverse);
+		}
+
+		// constant documentation mention it also accept any ICT transform
+		public bool Transform (ref CFRange range, CFString transform, bool reverse)
```
Comment (dalexsoto) [inline src/CoreFoundation/CFMutableString.cs]: What about this API to ease the ICT stuff? `NSString` and `CFString` are toll-free bridged.

```csharp
public bool Transform (ref CFRange range, string transform, bool reverse)
{
	str = null; // destroy any cached value
	var t = NSString.CreateNative (tagClass);
	var ret = CFStringTransform (Handle, ref range, t, reverse);
	NSString.ReleaseNative (t);
	return ret;
}
```

**Comment by dalexsoto** [comment]
Looks good so far 👍

File: src/CoreFoundation/CFMutableString.cs (hunk 1 above)
Comment (spouliot) [reply src/CoreFoundation/CFMutableString.cs]: I'd need to find real world examples first (for non-constants). It's never 100% clear if the comparison is done on strings (if should) or addresses (should not, but it happens).

Still the next version will be different and use`NSString` because that's how smart enums are exposed :-)

And from here it's not hard to use type`(NSString) String.Empty` if you have a `string` value.

File: src/CoreFoundation/CFMutableString.cs [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+
+		[DllImport (Constants.CoreFoundationLibrary, CharSet=CharSet.Unicode)]
+		static extern unsafe /* CFMutableStringRef* */ IntPtr CFStringCreateMutableWithExternalCharactersNoCopy (/* CFAllocatorRef* */ IntPtr alloc, string chars, nint numChars, nint capacity, /* CFAllocatorRef* */ IntPtr externalCharactersAllocator);
+
+		public CFMutableString (string theString, nint capacity)
+		{
+			if (theString == null)
+				throw new ArgumentNullException (nameof (theString));
+			
+			handle = CFStringCreateMutableWithExternalCharactersNoCopy (IntPtr.Zero, theString, theString.Length, capacity, IntPtr.Zero);
```
Comment (rolfbjarne) [inline src/CoreFoundation/CFMutableString.cs]: Is this really the native function you want to use? `This function permits you to create a CFMutableString object whose backing store is an external Unicode character buffer—that is, a buffer that you control (or can control) entirely.` It seems that if we want to use this function we'll have to copy the string to a buffer we control.

File: src/CoreFoundation/CFMutableString.cs [hunk 3]
```diff
@@ -0,0 +1,63 @@
...
+		}
+
+		[DllImport (Constants.CoreFoundationLibrary)]
+		static extern unsafe /* CFMutableStringRef* */ IntPtr CFStringCreateMutable (/* CFAllocatorRef* */ IntPtr alloc, nint maxLength);
+
+		public CFMutableString (nint maxLength)
```
Comment (rolfbjarne) [inline src/CoreFoundation/CFMutableString.cs]: Default value? `nint maxLength = 0`

File: src/CoreFoundation/CFMutableString.cs [hunk 4]
```diff
@@ -0,0 +1,63 @@
...
+		{
+			handle = CFStringCreateMutable (IntPtr.Zero, maxLength);
+		}
+
+		[DllImport (Constants.CoreFoundationLibrary)]
+		static extern unsafe /* CFMutableStringRef* */ IntPtr CFStringCreateMutableCopy (/* CFAllocatorRef* */ IntPtr alloc, nint maxLength, /* CFStringRef* */ IntPtr theString);
+
+		public CFMutableString (CFString theString, nint maxLength)
```
Comment (rolfbjarne) [inline src/CoreFoundation/CFMutableString.cs]: Default value? `nint maxLength = 0`

File: src/CoreFoundation/CFMutableString.cs [hunk 5]
```diff
@@ -0,0 +1,63 @@
...
+		{
+			handle = CFStringCreateMutableCopy (IntPtr.Zero, maxLength, theString.GetHandle ());
+		}
+
+		[DllImport (Constants.CoreFoundationLibrary)]
+		[return: MarshalAs (UnmanagedType.I1)]
+		static extern internal bool /* Boolean */ CFStringTransform (/* CFMutableStringRef* */ IntPtr @string, /* CFRange* */ ref CFRange range, /* CFStringRef* */ IntPtr transform, [MarshalAs (UnmanagedType.I1)] /* Boolean */ bool reverse);
+
+		public bool Transform (ref CFRange range, CFStringTransform transform, bool reverse)
```
Comment (rolfbjarne) [inline src/CoreFoundation/CFMutableString.cs]: `range` can be NULL, so maybe add an overload that doesn't take a `range`.

File: src/CoreFoundation/CFMutableString.cs [hunk 6]
```diff
@@ -0,0 +1,63 @@
...
+			
+			handle = CFStringCreateMutableWithExternalCharactersNoCopy (IntPtr.Zero, theString, theString.Length, capacity, IntPtr.Zero);
+		}
+
+		[DllImport (Constants.CoreFoundationLibrary)]
+		static extern unsafe /* CFMutableStringRef* */ IntPtr CFStringCreateMutable (/* CFAllocatorRef* */ IntPtr alloc, nint maxLength);
+
+		public CFMutableString (nint maxLength)
+		{
+			handle = CFStringCreateMutable (IntPtr.Zero, maxLength);
```
Comment (rolfbjarne) [inline src/CoreFoundation/CFMutableString.cs]: Maybe throw an exception if `handle` ends up being IntPtr.Zero?

File: src/CoreFoundation/CFMutableString.cs (hunk 3 above)
Comment (spouliot) [reply src/CoreFoundation/CFMutableString.cs]: I tried that first, but does not work with our `nint` :(
Maybe I'll use `int` and typecast, not ideal :|

File: src/CoreFoundation/CFMutableString.cs (hunk 5 above)
Comment (spouliot) [reply src/CoreFoundation/CFMutableString.cs]: good idea, I did not like a `ref CFRange?`

File: src/CoreFoundation/CFMutableString.cs (hunk 3 above)
Comment (rolfbjarne) [reply src/CoreFoundation/CFMutableString.cs]: You can do this: `nint maxLength = default (nint)`

File: src/CoreFoundation/CFMutableString.cs (hunk 6 above)
Comment (spouliot) [reply src/CoreFoundation/CFMutableString.cs]: I need to test what happens if you allocate a `nil` string and append to it
It's also something that we should do (and share) on the base `CFString` type

File: src/CoreFoundation/CFMutableString.cs (hunk 2 above)
Comment (spouliot) [reply src/CoreFoundation/CFMutableString.cs]: missed that, I'll see what's easier (make CFString mutable or copying)

File: src/CoreFoundation/CFMutableString.cs (hunk 6 above)
Comment (rolfbjarne) [reply src/CoreFoundation/CFMutableString.cs]: CFString should inherit from NativeObject, then we'd get much of this logic for free :smile:

File: src/CoreFoundation/CFMutableString.cs (hunk 3 above)
Comment (spouliot) [reply src/CoreFoundation/CFMutableString.cs]: oh, that's much nicer :) thanks!

File: src/CoreFoundation/CFMutableString.cs (hunk 6 above)
Comment (spouliot) [reply src/CoreFoundation/CFMutableString.cs]: yeah, it's getting farther and father away from the single missing API
but I guess it's a good time to bite the bullet on this one :)

File: src/CoreFoundation/CFMutableString.cs (hunk 5 above)
Comment (spouliot) [reply src/CoreFoundation/CFMutableString.cs]: done

File: src/CoreFoundation/CFMutableString.cs (hunk 1 above)
Comment (spouliot) [reply src/CoreFoundation/CFMutableString.cs]: done

File: src/CoreFoundation/CFMutableString.cs (hunk 6 above)
Comment (spouliot) [reply src/CoreFoundation/CFMutableString.cs]: done

File: src/CoreFoundation/CFMutableString.cs (hunk 4 above)
Comment (spouliot) [reply src/CoreFoundation/CFMutableString.cs]: done

File: src/CoreFoundation/CFMutableString.cs (hunk 2 above)
Comment (spouliot) [reply src/CoreFoundation/CFMutableString.cs]: removed

File: src/CoreFoundation/CFMutableString.cs [hunk 7]
```diff
@@ -0,0 +1,134 @@
...
+
+using System;
+using System.Runtime.InteropServices;
+
+using Foundation;
+using ObjCRuntime;
+
+namespace CoreFoundation {
+	
+	public class CFMutableString : CFString {
```
Comment (rolfbjarne) [inline src/CoreFoundation/CFMutableString.cs]: `sealed`?

File: src/CoreFoundation/CFMutableString.cs [hunk 8]
```diff
@@ -0,0 +1,134 @@
...
+
+		public CFMutableString (CFString theString, nint maxLength = default (nint))
+		{
+			if (theString == null)
+				throw new ArgumentNullException (nameof (theString));
+			Handle = CFStringCreateMutableCopy (IntPtr.Zero, maxLength, theString.GetHandle ());
+		}
+
+		[DllImport (Constants.CoreFoundationLibrary, CharSet=CharSet.Unicode)]
+		static extern unsafe void CFStringAppendCharacters (/* CFMutableStringRef* */ IntPtr theString, string chars, nint numChars);
```
Comment (rolfbjarne) [inline src/CoreFoundation/CFMutableString.cs]: `unsafe` isn't needed.

File: src/CoreFoundation/CFMutableString.cs [hunk 9]
```diff
@@ -0,0 +1,134 @@
...
+				throw new ArgumentNullException (nameof (@string));
+			Handle = CFStringCreateMutable (IntPtr.Zero, maxLength);
+			if (@string != null)
+				CFStringAppendCharacters (Handle, @string, @string.Length);
+		}
+
+		[DllImport (Constants.CoreFoundationLibrary)]
+		static extern unsafe /* CFMutableStringRef* */ IntPtr CFStringCreateMutableCopy (/* CFAllocatorRef* */ IntPtr alloc, nint maxLength, /* CFStringRef* */ IntPtr theString);
+
+		public CFMutableString (CFString theString, nint maxLength = default (nint))
```
Comment (rolfbjarne) [inline src/CoreFoundation/CFMutableString.cs]: `theString` -> `string` to match all the other API.

File: src/CoreFoundation/CFMutableString.cs [hunk 10]
```diff
@@ -0,0 +1,134 @@
...
+		[Preserve (Conditional = true)]
+		protected CFMutableString (IntPtr handle, bool owns)
+			: base (handle, owns)
+		{
+		}
+
+		[DllImport (Constants.CoreFoundationLibrary)]
+		static extern unsafe /* CFMutableStringRef* */ IntPtr CFStringCreateMutable (/* CFAllocatorRef* */ IntPtr alloc, nint maxLength);
+
+		public CFMutableString (string @string = "", nint maxLength = default (nint))
```
Comment (rolfbjarne) [inline src/CoreFoundation/CFMutableString.cs]: What happens if `maxLength < 0`?

File: src/CoreFoundation/CFMutableString.cs [hunk 11]
```diff
@@ -0,0 +1,134 @@
...
+
+		public bool Transform (ref CFRange range, NSString transform, bool reverse)
+		{
+			return Transform (ref range, transform.GetHandle (), reverse);
+		}
+
+		public bool Transform (ref CFRange range, string transform, bool reverse)
+		{
+			var t = NSString.CreateNative (transform);
+			var ret = Transform (ref range, t, reverse);
```
Comment (rolfbjarne) [inline src/CoreFoundation/CFMutableString.cs]:
```csharp
try {
    return Transform (ref range, t, reverse);
} finally {
    NSString.ReleaseNative (t);
}
```

File: src/CoreFoundation/CFMutableString.cs [hunk 12]
```diff
@@ -0,0 +1,134 @@
...
+		public bool Transform (NSString transform, bool reverse)
+		{
+			return Transform (transform.GetHandle (), reverse);
+		}
+
+		public bool Transform (string transform, bool reverse)
+		{
+			var t = NSString.CreateNative (transform);
+			var ret = Transform (t, reverse);
+			NSString.ReleaseNative (t);
```
Comment (rolfbjarne) [inline src/CoreFoundation/CFMutableString.cs]: Same: try/finally.

File: tests/monotouch-test/CoreFoundation/MutableString.cs [hunk 13]
```diff
@@ -0,0 +1,124 @@
...
+				Assert.That (s.Handle, Is.Not.EqualTo (IntPtr.Zero), "Handle");
+				Assert.That (s.ToString (), Is.Empty, "ToString");
+			}
+		}
+
+		[Test]
+		public void CreateString1 ()
+		{
+			Assert.Throws<ArgumentNullException> (() => new CFMutableString ((string)null), "null");
+			using (var s = new CFMutableString ("bonjour!")) {
```
Comment (rolfbjarne) [inline tests/monotouch-test/CoreFoundation/MutableString.cs]: More complex test strings would be nice too: emojis, bidi characters, unicode characters with multiple code points, etc.

File: src/CoreFoundation/CFMutableString.cs [hunk 14]
```diff
@@ -0,0 +1,134 @@
...
+using System.Runtime.InteropServices;
+
+using Foundation;
+using ObjCRuntime;
+
+namespace CoreFoundation {
+	
+	public class CFMutableString : CFString {
+
+		public CFMutableString (IntPtr handle)
```
Comment (rolfbjarne) [inline src/CoreFoundation/CFMutableString.cs]: `internal`? I know the CFString version is public, but that doesn't mean we have to add new public ones.

File: src/CoreFoundation/CFMutableString.cs (hunk 9 above)
Comment (spouliot) [reply src/CoreFoundation/CFMutableString.cs]: That's the same names as the native API, which makes it easier to understand (when the documentation merger is used).

File: src/CoreFoundation/CFMutableString.cs (hunk 7 above)
Comment (spouliot) [reply src/CoreFoundation/CFMutableString.cs]: why ? it's not common in our bindings (for direct, public bindings) and the linker can do that optimization.

File: src/CoreFoundation/CFMutableString.cs (hunk 10 above)
Comment (spouliot) [reply src/CoreFoundation/CFMutableString.cs]: worth a test

File: src/CoreFoundation/CFMutableString.cs (hunk 14 above)
Comment (spouliot) [reply src/CoreFoundation/CFMutableString.cs]: changing to `protected` like the 2 argument `.ctor`

The swift bindings do not seems to make to limit subclassing this type - and I don't see why we should impose our own arbitrary limits (there's enough _non-arbitrary _ of them already)

File: src/CoreFoundation/CFMutableString.cs (hunk 11 above)
Comment (spouliot) [reply src/CoreFoundation/CFMutableString.cs]: good point, `Transform` can now throw

File: tests/monotouch-test/CoreFoundation/MutableString.cs (hunk 13 above)
Comment (spouliot) [reply tests/monotouch-test/CoreFoundation/MutableString.cs]: Tests are to validate the bindings, not the API. That historically caused a lot of troubles (false positives) when run on devices with different OS versions or installed/default languages.

Is there something specific you think would cause problems ?

**Comment by spouliot** [comment]
bot/network issue
```
09:05:37  > git submodule update --init --recursive external/mono # timeout=30
09:35:37 ERROR: Timeout after 30 minutes
09:35:37 Also:   hudson.remoting.Channel$CallSiteStackTrace: Remote call to JNLP4-connect connection from 167.220.149.94/167.220.149.94:15339
```

**Comment by spouliot** [comment]
build

File: tests/monotouch-test/CoreFoundation/MutableString.cs (hunk 13 above)
Comment (rolfbjarne) [reply tests/monotouch-test/CoreFoundation/MutableString.cs]: Yes, I'd like to confirm if Mono's understanding of Unicode is the same as Apple's, to ensure that a roundtrip string -> CFString -> string doesn't end up with something corrupted.

For instance: `string.Length` is "the number of Char objects in this instance, not the number of Unicode characters.", while `CFStringAppendCharacters` takes `numChars:
The number of Unicode characters in chars`. I don't trust documentation, but it seems there is a possibility that `CFStringAppendCharacters` can go off into random memory looking for more characters than we actually provided.

Unicode is hard 😒

File: src/CoreFoundation/CFMutableString.cs (hunk 7 above)
Comment (rolfbjarne) [reply src/CoreFoundation/CFMutableString.cs]: It seems that subclassing CFMutableString is something nobody should ever do, and besides it's always possible unseal it later if we want to.

Comment (spouliot) [reply src/CoreFoundation/CFMutableString.cs]: Swift does not seal it and I can't see any positive outcome out of sealing it. At best it does nothing (linker will seal it), at worse someone will need that (for good or bad reasons) and it will takes weeks/months to get a fix out.

**Comment by spouliot** [comment]
@monojenkins backport to d16-1
